### PR TITLE
Introduce v1beta1 API version

### PR DESCRIPTION
**Pull request by csrwng**
**What this PR does / why we need it**:
Adds code to convert from v1alpha1 to v1beta1
Updates code to use the new version

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Ref https://issues.redhat.com/browse/HOSTEDCP-512

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Review by imain — COMMENTED** [review]
Looks like a very nice cleanup!

File: api/v1beta1/endpointservice_types.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	// EndpointServiceName is the name of the Endpoint Service created in the
+	// management VPC
+	// +optional
+	EndpointServiceName string `json:"endpointServiceName,omitempty"`
+
+	// EndpointID is the ID of the Endpoint created in the guest VPC
+	// +optional
+	EndpointID string `json:"endpointID,omitempty"`
+
+	// Deprecated: Use DNSNames instead
```
Comment (enxebre) [inline api/v1beta1/endpointservice_types.go:50]: Aren't we dropping all deprecated fields?

File: Makefile [hunk 2]
```diff
@@ -147,7 +147,7 @@ app-sre-saas-template: hypershift
 # Run tests
 .PHONY: test
 test:
-	$(GO) test -race -count=25 -timeout=20m ./... -coverprofile cover.out
+	$(GO) test -race -count=25 -timeout=30m ./... -coverprofile cover.out
```
[the remark this answers is not in the document]
Comment (csrwng) [reply Makefile:150]: The reason for the larger timeout is the fuzzer conversion tests. For whatever reason, we can't contain them in CI in less than 20 min. They run much faster on your laptop.

File: api/v1beta1/endpointservice_types.go (hunk 1 above)
Comment (csrwng) [reply api/v1beta1/endpointservice_types.go:50]: yes, need to remove

**Comment by csrwng** [comment]
@enxebre updated the CLI code to work with v1beta1 by default but to fallback to v1alpha1 if v1beta1 is not present. 
This should be ready for another review, all remaining items have been addressed.

File: api/v1beta1/hostedcluster_types.go [hunk 3]
```diff
@@ -1658,6 +1665,7 @@ const (
 	HostedControlPlaneComponentsUnavailableReason = "ComponentsUnavailable"
 	KubeconfigWaitingForCreateReason              = "KubeconfigWaitingForCreate"
 	ClusterVersionStatusUnknownReason             = "ClusterVersionStatusUnknown"
+	FromClusterVersionReason                      = "FromClusterVersion"
```
Comment (enxebre) [inline api/v1beta1/hostedcluster_types.go:1668]: does this change sneaked in?

File: api/v1alpha1/zz_conversion.go [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+		return err
+	}
+	return nil
+}
+
+func fixupHostedClusterAfterConversion(hc *HostedCluster) error {
+	if hc.Spec.Platform.AWS != nil {
+		populatedDeprecatedAWSRoles(hc.Spec.Platform.AWS)
+	}
+	populateDeprecatedNetworkingFields(&hc.Spec.Networking)
```
Comment (enxebre) [inline api/v1alpha1/zz_conversion.go:124]: why do we need to populate deprecated fields? v1alpha1 has all the newest ones right?

**Comment by enxebre** [comment]
why do we need the client fallback vs asking folks to install v1beta1 or stay with old cli?

File: api/v1beta1/hostedcluster_types.go [hunk 5]
```diff
@@ -117,6 +117,13 @@ const (
 	// removed when deleting the corresponding HostedCluster. If set to "true", resources created on the cloud provider during the life
 	// of the cluster will be removed, including image registry storage, ingress dns records, load balancers, and persistent storage.
 	CleanupCloudResourcesAnnotation = "hypershift.openshift.io/cleanup-cloud-resources"
+
+	// ResourceRequestOverrideAnnotationPrefix is a prefix for an annotation to override resource requests for a particular deployment/container
+	// in a hosted control plane. The format of the annotation is:
+	// resource-request-override.hypershift.openshift.io/[deployment-name].[container-name]: [resource-type-1]=[value1],[resource-type-2]=[value2],...
```
[the remark this answers is not in the document]
Comment (csrwng) [reply api/v1beta1/hostedcluster_types.go:123]: These are copied from the v1alpha1 package so we can have the same in v1beta1 which is now what we'll use going forward. The thinking is that once we switch to v1beta1 we will only add new constants to the v1beta1 package and leave the v1alpha1 alone.

**Comment by csrwng** [comment]
> why do we need the client fallback vs asking folks to install v1beta1 or stay with old cli?

As mentioned earlier, it's to avoid UX issues while we transition to the new API version. It is pretty easy to remove the fallback once our consumers have fully transitioned to v1beta1.

**Pull request merged**